tool/tctl/common/bots_command.go

```diff
@@ -238,9 +283,156 @@ func (c *BotsCommand) AddBot(ctx context.Context, client auth.ClientI) error {
+		log.Warning("No roles specified. The bot will not be able to produce outputs until a role is added to the bot.")
+	}
+	var token types.ProvisionToken
+	var err error
+	if c.tokenID == "" {
+		// If there's no token specified, generate one
+		tokenName, err := utils.CryptoRandomHex(16)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		ttl := c.tokenTTL
+		if ttl == 0 {
+			ttl = defaults.DefaultBotJoinTTL
+		}
+		tokenSpec := types.ProvisionTokenSpecV2{
+			Roles:      types.SystemRoles{types.RoleBot},
+			JoinMethod: types.JoinMethodToken,
+			BotName:    c.botName,
+		}
+		token, err = types.NewProvisionTokenFromSpec(tokenName, time.Now().Add(ttl), tokenSpec)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if err := client.UpsertToken(ctx, token); err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		// If there is, check the token matches the potential bot
+		token, err = client.GetToken(ctx, c.tokenID)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return trace.NotFound("token with name %q not found, create the token or do not set TokenName: %v",
+					c.tokenID, err)
+			}
+			return trace.Wrap(err)
+		}
+		if !token.GetRoles().Include(types.RoleBot) {
+			return trace.BadParameter("token %q is not valid for role %q",
+				c.tokenID, types.RoleBot)
+		}
+		if token.GetBotName() != c.botName {
+			return trace.BadParameter("token %q is valid for bot with name %q, not %q",
+				c.tokenID, token.GetBotName(), c.botName)
+		}
+	}
+
+	bot := &machineidv1pb.Bot{
+		Metadata: &headerv1.Metadata{
+			Name: c.botName,
+		},
+		Spec: &machineidv1pb.BotSpec{
+			Roles: roles,
+			Traits: []*machineidv1pb.Trait{
+				{
+					Name:   constants.TraitLogins,
+					Values: flattenSlice(c.allowedLogins),
+				},
+			},
+		},
+	}
+
+	bot, err = client.BotServiceClient().CreateBot(ctx, &machineidv1pb.CreateBotRequest{
+		Bot: bot,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if c.format == teleport.JSON {
+		tokenTTL := time.Duration(0)
+		if exp := token.Expiry(); !exp.IsZero() {
+			tokenTTL = time.Until(exp)
+		}
+		// This struct is equivalent to a legacy bit of JSON we used to output
+		// when we called an older RPC. We've preserved it here to avoid
+		// breaking customer scripts.
+		response := struct {
+			UserName string        `json:"user_name"`
+			RoleName string        `json:"role_name"`
+			TokenID  string        `json:"token_id"`
+			TokenTTL time.Duration `json:"token_ttl"`
+		}{
+			UserName: bot.Status.UserName,
+			RoleName: bot.Status.RoleName,
+			TokenID:  token.GetName(),
+			TokenTTL: tokenTTL,
+		}
+		out, err := json.MarshalIndent(response, "", "  ")
+		if err != nil {
+			return trace.Wrap(err, "failed to marshal CreateBot response")
+		}
+
+		fmt.Println(string(out))
+		return nil
+	}
+
+	proxies, err := client.GetProxies()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(proxies) == 0 {
+		return trace.Errorf("This cluster does not have any proxy servers running.")
```

Comment:
Edgiest of edge cases, but at this point the bot has been created, so getting an error might confuse the customer about the state of their cluster. Consider acknowledging that the bot was created in this error message.

Reply:
Yeah, although this is the old message as well, I'll go ahead and update it since I agree this is unclear.

Reply:
3b27833c7b